test/unittests/test_zx_functionality.cpp

```diff
@@ -0,0 +1,202 @@
+    EXPECT_EQ(diag.getEdge(6, 1).value().type, zx::EdgeType::Simple);
+    EXPECT_EQ(diag.getEdge(3, 6).value().type, zx::EdgeType::Simple);
+    EXPECT_EQ(diag.getEdge(4, 5).value().type, zx::EdgeType::Simple);
+    EXPECT_EQ(diag.getEdge(5, 2).value().type, zx::EdgeType::Simple);
+
+    EXPECT_EQ(diag.getVData(0).value().type, zx::VertexType::Boundary);
+    EXPECT_EQ(diag.getVData(1).value().type, zx::VertexType::Boundary);
+    EXPECT_EQ(diag.getVData(4).value().type, zx::VertexType::Z);
+    EXPECT_EQ(diag.getVData(5).value().type, zx::VertexType::Z);
+    EXPECT_EQ(diag.getVData(6).value().type, zx::VertexType::X);
+    EXPECT_EQ(diag.getVData(2).value().type, zx::VertexType::Boundary);
+    EXPECT_EQ(diag.getVData(3).value().type, zx::VertexType::Boundary);
+
+    for (std::size_t i = 0; i < diag.getNVertices(); i++)
+        EXPECT_TRUE(diag.getVData(i).value().phase.isZero());
+}
+
+TEST_F(ZXDiagramTest, complex_circuit) {
+    std::stringstream ss{};
+    ss << "// i 1 0 2\n"
+       << "// o 0 1 2\n"
+       << "OPENQASM 2.0;"
+       << "include \"qelib1.inc\";"
+       << "qreg q[3];"
+       << "h q[0];"
+       << "cx q[0],q[1];"
+       << "z q[1];"
+       << "x q[2];"
+       << "y q[0];"
+       << "rx(pi/4) q[0];"
+       << "rz(0.1) q[1];"
+       << "p(0.1) q[1];"
+       << "ry(pi/4) q[2];"
+       << "t q[0];"
+       << "s q[2];"
+       << "u2(pi/4, pi/4) q[1];"
+       << "u3(pi/4, pi/4, pi/4) q[2];"
+       << "barrier q[0],q[1],q[2];"
+       << "swap q[0],q[1];"
+       << "cz q[1],q[2];"
+       << "cp(pi/4) q[0],q[1];"
+       << "ccx q[0],q[1],q[2];"
+       << "ccz q[1],q[2],q[0];"
+       << "cp(pi/2) q[0], q[1];"
+       << "cp(pi/4) q[0], q[1];"
+       << "cp(-pi/4) q[0], q[1];"
+       << "cp(-pi/2) q[0], q[1];"
+       << "ccz q[1],q[2],q[0];"
+       << "ccx q[0],q[1],q[2];"
+       << "cp(-pi/4) q[0],q[1];"
+       << "cz q[1],q[2];"
+       << "cx q[1],q[0];"
+       << "cx q[0],q[1];"
+       << "cx q[1],q[0];"
+       << "u3(-pi/4,-pi/4,-pi/4) q[2];"
+       << "u2(-5*pi/4,3*pi/4) q[1];"
+       << "sdg q[2];"
+       << "tdg q[0];"
+       << "ry(-pi/4) q[2];"
+       << "p(-0.1) q[1];"
+       << "rz(-0.1) q[1];"
+       << "rx(-pi/4) q[0];"
+       << "y q[0];"
+       << "x q[2];"
+       << "z q[1];"
+       << "cx q[0],q[1];"
+       << "h q[0];"
+       << std::endl;
+    qc.import(ss, qc::OpenQASM);
+
+    EXPECT_TRUE(zx::FunctionalityConstruction::transformableToZX(&qc));
+    zx::ZXDiagram diag = zx::FunctionalityConstruction::buildFunctionality(&qc);
+    zx::fullReduce(diag);
+    EXPECT_EQ(diag.getNVertices(), 6);
+    EXPECT_EQ(diag.getNEdges(), 3);
+    EXPECT_TRUE(diag.isIdentity());
+}
+
+TEST_F(ZXDiagramTest, Phase) {
+    qc = qc::QuantumComputation(2);
+    qc.phase(0, zx::PI / 4);
+    qc.phase(0, dd::Control{1, dd::Control::Type::pos}, zx::PI / 4);
+    qc.phase(0, dd::Control{1, dd::Control::Type::pos}, -zx::PI / 4);
```

Comment:
you can avoid typing these controls in such a verbose fashion by 
```
using namespace dd::literals;
```
and then using `1_pc` for positive controls and `1_nc` for negative controls.